### PR TITLE
Add functionality to not combine abutting intervals

### PR DESCRIPTION
**Pull request by tmelman**
### Description

The behavior up until now in `IntervalListTools` for dealing with combining interval lists has been to combine abutting and overlapping interval lists. However, GATK internally often keeps abutting intervals separate, as detailed in this issue: https://github.com/broadinstitute/picard/issues/1499. 

Here I've added a functionality to create the interval list as requested, keeping abutting interval lists separate. 

Changes made: 
* Added an option for `DONT_COMBINE_ABUTTING` to `IntervalListTools`. Default is `false` to keep current behavior the same. 
* When `DONT_COMBINE_ABUTTING=true` and `UNIQUE=true` and there is no action, this will create interval lists merging ONLY overlapping intervals.
* Throw an error when `DONT_COMBINE_ABUTTING=true` but `ACTION=UNION`, as this is unsupported.
* Added tests and test data. 

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/util/IntervalListTools.java [hunk 1]
```diff
@@ -402,15 +405,18 @@ protected int doWork() {
...
+            SORT = false; // no need to sort, since uniqued() output will be sorted by definition.
             UNIQUE = true;
         }
 
         final IntervalList possiblySortedResult = SORT ? result.sorted() : result;
         final IntervalList possiblyInvertedResult = INVERT ? IntervalList.invert(possiblySortedResult) : possiblySortedResult;
 
         //only get unique if this has been asked unless inverting (since the invert will return a unique list)
-        List<Interval> finalIntervals = UNIQUE ? possiblyInvertedResult.uniqued().getIntervals() : possiblyInvertedResult.getIntervals();
+//        List<Interval> finalIntervals = UNIQUE ? possiblyInvertedResult.uniqued().getIntervals() : possiblyInvertedResult.getIntervals();
```
Comment (kachulis) [inline src/main/java/picard/util/IntervalListTools.java:416]: remove commented out old code

File: src/test/java/picard/util/IntervalListToolsTest.java [hunk 2]
```diff
@@ -406,4 +409,63 @@ public void testLargeScatters() throws IOException {
 
         Assert.assertEquals(gather, original);
     }
+
+    @Test(timeOut = 40_000)
```
Comment (kachulis) [inline src/test/java/picard/util/IntervalListToolsTest.java:413]: don't need the `timeOut` since you're testing on very small files.

File: src/main/java/picard/util/IntervalListTools.java [hunk 3]
```diff
@@ -513,6 +519,9 @@ private IntervalList openIntervalLists(final List<File> files, BinaryOperator<In
         if (COUNT_OUTPUT != null && OUTPUT_VALUE == Output.NONE) {
             errorMsgs.add("COUNT_OUTPUT was provided but OUTPUT_VALUE is set to NONE.");
         }
+        if (ACTION == Action.UNION && DONT_COMBINE_ABUTTING) {
+            errorMsgs.add("ACTION=UNION action combines abutting intervals by default, so it is incompatible with DONT_COMBINE_ABUTTING=true.");
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:523]: I'm not sure I understand why this is inherently incompatible with UNION. I get that _currently_ the htsjdk `uniqued` method combines abbutting intervals, but I don't see why it has to be that way....it seems plausible, to work around it by using the lower-level `getUniqueIntervals` directly.

File: src/test/java/picard/util/IntervalListToolsTest.java [hunk 4]
```diff
@@ -406,4 +409,63 @@ public void testLargeScatters() throws IOException {
...
+    public void testCombineAbuttingIntervals() throws IOException {
+        // Test the default behavior of UNION, which is to combine abutting and overlapping intervals.
+        //gather
+        final File ilOut = File.createTempFile("IntervalListTools", ".interval_list");
+        ilOut.deleteOnExit();
+        {
```
Comment (kachulis) [inline src/test/java/picard/util/IntervalListToolsTest.java:419]: unneeded brackets (maybe this used to be inside a try block?)

**Review by yfarjoun — CHANGES_REQUESTED** [review]
Thanks for writing up all these nice tests!

I think you should be able to enable this to work with UNION as well...I have faith in you!

File: src/test/java/picard/util/IntervalListToolsTest.java [hunk 5]
```diff
@@ -406,4 +409,63 @@ public void testLargeScatters() throws IOException {
...
+            args.add("ACTION=UNION"); // Note: ACTION=UNION is equivalent to UNIQUE=true and DONT_COMBINE_ABUTTING=false, or defaults
+            Assert.assertEquals(runPicardCommandLine(args), 0);
+        }
+        final IntervalList gather = IntervalList.fromFile(ilOut);
+        final IntervalList original = IntervalList.fromFile(abutting_combined);
+
+        Assert.assertEquals(gather, original); // equal to expected output
+    }
+    @Test(timeOut = 40_000)
+    public void testDontCombineAbuttingIntervals() throws IOException {
```
Comment (kachulis) [inline src/test/java/picard/util/IntervalListToolsTest.java:432]: Would like to see some tests of the other actions as well.  And I think you can combine this with `testCombineAbuttingIntervals` and use a DataProvider

**Review by kachulis — COMMENTED** [review]
nice!  a few comments, and I think the tests could be condensed and expanded with a DataProvider.

File: src/main/java/picard/util/IntervalListTools.java (hunk 3 above)
Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java:533]: There is an additional wrinkle, that `getUniqueIntervals` returns `List<Intervals>` whereas `uniqued()` returns `IntervalList`, and there is no constructor that takes in a `List<Intervals>` and turns it into an `IntervalList`. I can see if there's a way to do this with `addAll()` and removing the old intervals. If you have any ideas let me know.

File: src/test/java/picard/util/IntervalListToolsTest.java (hunk 4 above)
Comment (tmelman) [reply src/test/java/picard/util/IntervalListToolsTest.java:419]: I was wondering about this, I copied and pasted from another test. Happy to remove! :)

File: src/main/java/picard/util/IntervalListTools.java (hunk 3 above)
Comment (yfarjoun) [reply src/main/java/picard/util/IntervalListTools.java:533]: you can replicate the htsjdk code to make a private method that also allows you to state "mergeAbutting", right? eventually perhaps someone will push that down to htsdjk.

Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java]: Fixed @yfarjoun

File: src/test/java/picard/util/IntervalListToolsTest.java (hunk 5 above)
Comment (tmelman) [reply src/test/java/picard/util/IntervalListToolsTest.java]: Fixed @kachulis

File: src/main/java/picard/util/IntervalListTools.java [hunk 6]
```diff
@@ -501,6 +514,22 @@ private IntervalList openIntervalLists(final List<File> files, BinaryOperator<In
                 .orElse(null);
     }
 
+    private static IntervalList uniqued(IntervalList non_unique, boolean merge_abutting) {
```
Comment (kachulis) [inline src/main/java/picard/util/IntervalListTools.java:517]: why does this need to return an IntervalList? can't you let it return a list of intervals instead?  You immediately call `getIntervals()` on it the only place I see it used, so it seems like a lot of unneeded converting back and forth between lists of intervals and interval lists.

Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java:517]: Because I'm reproducing the functionality of "uniqued" in htsjdk's IntervalList. The idea is to have a "uniqued" method that we can one day add to their code base in a PR. Alternatively they have a "getUniqueIntervals()" method that rerturns a List of Intervals. I could use that but this makes for cleaner code with all the defaults. 

tl;dr: this was purely a design decision, I don't have a strong preference on which way to do it, and I'm happy to change it to according to your preferences, as you will also be interacting with it.

Comment (kachulis) [reply src/main/java/picard/util/IntervalListTools.java:517]: I see, that seems reasonable then

**Review by kachulis — APPROVED** [review]
looks good!

File: testdata/picard/util/abutting.interval_list [hunk 7]
```diff
@@ -0,0 +1,6 @@
+@HD	VN:1.6	SO:coordinate
+@SQ	SN:1	LN:249250621	AS:GRCh37	UR:http://www.broadinstitute.org/ftp/pub/seq/references/Homo_sapiens_assembly19.fasta	M5:1b22b98cdeb4a9304cb5d48026a85128	SP:Homo Sapiens
+1	30000	30001	+	target_1
+1	30002	30003	+	target_2
```
Comment (yfarjoun) [inline testdata/picard/util/abutting.interval_list:4]: perhaps use the name of the interval to comment about the properties of each interval:

e.g 

```
1	30000	30001	+	target_1_abutting_the_next
1	30002	30003	+	target_2_abutting_the_previous_but_not_the_next
1	30005	30007	+	target_3_overlapping_the_next_but_not_abutting_the_previous
1	30006	30008	+	target_4_overlapping_the_previous
```

File: testdata/picard/util/abutting_combined.interval_list [hunk 8]
```diff
@@ -0,0 +1,4 @@
+@HD	VN:1.6	SO:coordinate
+@SQ	SN:1	LN:249250621	AS:GRCh37	UR:http://www.broadinstitute.org/ftp/pub/seq/references/Homo_sapiens_assembly19.fasta	M5:1b22b98cdeb4a9304cb5d48026a85128	SP:Homo Sapiens
+1	30000	30003	+	target_1
```
Comment (yfarjoun) [inline testdata/picard/util/abutting_combined.interval_list:3]: if you name the intervals in a meaningful way, you can also have the tests "CONTATENATE=true" and the resulting names here will show which intervals got merged.

File: src/main/java/picard/util/IntervalListTools.java [hunk 9]
```diff
@@ -495,6 +514,22 @@ private IntervalList openIntervalLists(final List<File> files, BinaryOperator<In
...
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:517]: I'd add a "//TODO:'' item to move this to htsjdk and use that once we can. (possibly adding issues that can be referenced)

File: src/main/java/picard/util/IntervalListTools.java [hunk 10]
```diff
@@ -402,15 +416,20 @@ protected int doWork() {
...
             UNIQUE = true;
         }
 
         final IntervalList possiblySortedResult = SORT ? result.sorted() : result;
         final IntervalList possiblyInvertedResult = INVERT ? IntervalList.invert(possiblySortedResult) : possiblySortedResult;
 
-        //only get unique if this has been asked unless inverting (since the invert will return a unique list)
-        List<Interval> finalIntervals = UNIQUE ? possiblyInvertedResult.uniqued().getIntervals() : possiblyInvertedResult.getIntervals();
+        //only get unique if this has been asked OR if action is UNION, unless inverting (since the invert will return a unique list)
+        List<Interval> finalIntervals = UNIQUE? IntervalListTools.uniqued(possiblyInvertedResult, !DONT_MERGE_ABUTTING).getIntervals() : possiblyInvertedResult.getIntervals();
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:432]:
```suggestion
        List<Interval> finalIntervals = UNIQUE ? uniqued(possiblyInvertedResult, !DONT_MERGE_ABUTTING).getIntervals() : possiblyInvertedResult.getIntervals();
```

File: src/main/java/picard/util/IntervalListTools.java (hunk 9 above)
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:517]: We tend to use camelCase for variables and methods
```suggestion
    private static IntervalList uniqued(IntervalList nonUnique, boolean mergeAbutting) {
```

File: src/main/java/picard/util/IntervalListTools.java [hunk 11]
```diff
@@ -495,6 +514,22 @@ private IntervalList openIntervalLists(final List<File> files, BinaryOperator<In
...
 
+    private static IntervalList uniqued(IntervalList non_unique, boolean merge_abutting) {
+        // A subroutine to replace htsjdk's IntervalList "unique()" method, which combines abutting intervals by default.
+        // Returns an IntervalList.
+        // Inputs:
+        //  - non_unique: an IntervalList to be operated on
+        //  - merge_abutting: a boolean to combine abutting intervals (if true) or not (if false).
+        // Outputs:
+        //  - an IntervalList that is uniqued with overlapping intervals merged, and abutting intervals handled per the boolean flag.
+        final boolean CONCATENATE_NAMES = true;
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:525]: why are these in variables? you can just put `true,false` in the method call, no?

File: src/main/java/picard/util/IntervalListTools.java [hunk 12]
```diff
@@ -402,15 +416,20 @@ protected int doWork() {
         final IntervalList result = ACTION.act(lists, secondLists);
 
         if (INVERT) {
-            SORT = false; // no need to sort, since return will be sorted by definition.
+            SORT = false; // no need to sort, since uniqued() output will be sorted by definition.
+            UNIQUE = true;
+        }
+
+        if (ACTION == Action.UNION) { // UNION is basically Action.CONCAT with SORT and UNIQUE
+            SORT = true;
```
Comment (yfarjoun) [inline src/main/java/picard/util/IntervalListTools.java:424]: if UNIQUE sorts by default, shouldn't this be `SORT=false`?

**Review by yfarjoun — CHANGES_REQUESTED** [review]
Thanks for the srcond round!

a few more small nit-picks!

File: src/main/java/picard/util/IntervalListTools.java (hunk 11 above)
Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java:525]: I definitely can, it just makes for more readable code. Purely design decision, happy to change it

File: src/main/java/picard/util/IntervalListTools.java (hunk 12 above)
Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java:424]: per the previous implementation, UNION assumes sort and unique are true. so if action is union we have to reproduce that behavior.

File: src/main/java/picard/util/IntervalListTools.java (hunk 11 above)
Comment (yfarjoun) [reply src/main/java/picard/util/IntervalListTools.java:525]: OK. So in that case, just change the variable names to camelCase

File: src/main/java/picard/util/IntervalListTools.java (hunk 12 above)
Comment (yfarjoun) [reply src/main/java/picard/util/IntervalListTools.java:424]: I guess I'm confused by why the INVERT==true branch claims (in the comments) that theres no need to sort, and here there seems to be a need to sort again.

Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java:424]: theres a bit of a loop of flags and actions overlapping lol… but I don't know what "UNION" would be with "UNIQUE" false.

File: src/main/java/picard/util/IntervalListTools.java (hunk 11 above)
Comment (tmelman) [reply src/main/java/picard/util/IntervalListTools.java]: done 👍 :)

**Comment by tmelman** [comment]
@yfarjoun round 2 changes resolved, can you take a look?

**Review by yfarjoun — APPROVED** [review]

**Comment by yfarjoun** [comment]
merge away! Thanks for this!!

🎉

**Comment by yfarjoun** [comment]
you can delete your branch now @tmelman
